File: <100/061/061.py
import itertools


def triangle_prod(n: int) -> int:
    return n * (n+1) // 2


def square_prod(n: int) -> int:
    return n * n


def pentagonal_prod(n: int) -> int:
    return n * (3*n-1) // 2


def hexagonal_prod(n: int) -> int:
    return n * (2*n-1)


def heptagonal_prod(n: int) -> int:
    return n * (5*n-3) // 2


def octagonal_prod(n: int) -> int:
    return n * (3*n-2)


def eval_geo_vect(f: callable) -> list[int]:
    return [f(n) for n in range(1,220) if 1000 <= f(n) and f(n) <= 9999]


def eval_geo_vects() -> list[list[int]]:
    fs = [triangle_prod, square_prod, pentagonal_prod, hexagonal_prod, heptagonal_prod, octagonal_prod]
    results = map(lambda f: eval_geo_vect(f), fs)
    return list(results)


def is_chain(pre: int, now: int) -> bool:
    if pre == 0:
        return True
    tail = pre % 100
    head = now // 100
    return tail == head


def dfs(v_idx: list[int],
        geo_vects: list[list[int]],
        chains: list[list[int]],
        chain: list[int],
        n: int = 0,
        pre: int = 0,
        ) -> None:
    if n == len(v_idx):
        if is_chain(chain[-1], chain[0]):
            tmp = [a for a in chain]
            chains.append(tmp)
        return
    
    gv = filter(lambda x: is_chain(pre, x), geo_vects[v_idx[n]])
    for a in gv:
        chain.append(a)
        dfs(v_idx, geo_vects, chains, chain, n+1, a)
        chain.pop(-1)


def main() -> int:
    geo_vects = eval_geo_vects()
    v_idx_ = [j for j in range(1, len(geo_vects))]
    chains = []

    perms = itertools.permutations(v_idx_)
    for perm in perms:
        v_idx = [0] + list(perm)
        chain  = []
        dfs(v_idx, geo_vects, chains, chain)

    answer = sum(sum(chain) for chain in chains)
    return answer


if __name__ == "__main__":
    answer = main()
    print(answer)

File: <100/061/061.jl
using Combinatorics


triangle_prod(n :: Int) :: Int = n * (n+1) ÷ 2
square_prod(n :: Int) :: Int = n * n
pentagonal_prod(n :: Int) :: Int = n * (3*n-1) ÷ 2
hexagonal_prod(n :: Int) :: Int = n * (2*n-1)
heptagonal_prod(n :: Int) :: Int = n * (5*n-3) ÷ 2
octagonal_prod(n :: Int) :: Int = n * (3*n-2)


function eval_get_vect(f :: Function)
  return [f(n) for n ∈ 1:200 if 1000 ≤ f(n) && f(n) ≤ 9999]
end


function eval_geo_vects() :: Vector{Vector{Int}}
  fs = [triangle_prod, square_prod, pentagonal_prod, hexagonal_prod, heptagonal_prod, octagonal_prod]
  return map(f -> eval_get_vect(f), fs)
end


function is_chain(pre :: Int, now :: Int) :: Bool
  if pre == 0
    return true
  end
  tail = pre % 100
  head = now ÷ 100
  return tail == head
end



function dfs(v_idx :: Vector{Int},
             geo_vects :: Vector{Vector{Int}},
             chains :: Vector{Vector{Int}},
             chain :: Vector{Int},
             n :: Int = 1,
             pre :: Int = 0
) :: Nothing
  if n == length(v_idx) + 1
    if is_chain(chain[end], chain[begin])
      tmp = [a for a ∈ chain]
      push!(chains, tmp)
    end
    return nothing
  end

  gv = filter(x -> is_chain(pre, x), geo_vects[v_idx[n]])
  for a ∈ gv
    push!(chain, a)
    dfs(v_idx, geo_vects, chains, chain, n+1, a)
    pop!(chain)
  end

  return nothing
end


function main() :: Nothing
  geo_vects = eval_geo_vects()
  v_idx_ :: Vector{Int} = [j for j ∈ 2:length(geo_vects)]
  chains :: Vector{Vector{Int}} = []

  perms = permutations(v_idx_)
  for v_idx ∈ perms
    pushfirst!(v_idx, 1)
    chain :: Vector{Int} = []
    dfs(v_idx, geo_vects, chains, chain)
  end

  answer = sum([sum(chain) for chain ∈ chains])
  println(answer)
  return nothing
end


@time main()
ß